code/object/collideshipweapon.cpp

```diff
@@ -66,7 +66,7 @@ static void update_danger_weapon(object *pship_obj, object *weapon_obj)
  * Deal with weapon-ship hit stuff.  
  * Separated from check_collision routine below because of multiplayer reasons.
  */
-static void ship_weapon_do_hit_stuff(object *pship_obj, object *weapon_obj, vec3d *world_hitpos, vec3d *hitpos, int quadrant_num, int submodel_num, vec3d /*not a pointer intentionaly*/ hit_dir)
+static void ship_weapon_do_hit_stuff(object *pship_obj, object *weapon_obj, vec3d *world_hitpos, vec3d *hitpos, int quadrant_num, int submodel_num, const vec3d *hit_dir)
```

Comment:
If it used to not be a pointer intentionally, there needs to be *some* kind of comment explaining why changing it won't cause problems.

Reply:
That was what I alluded to here:
> Another vec3d copy can be replaced with a vec3d pointer since there is no longer a need to keep a local copy.

To elaborate, at one point in time `ship_weapon_do_hit_stuff()` modified `hit_dir`, so the parameter was copied in order to avoid modifying the vec3d from the calling function.  Later on the code that modified `hit_dir` was removed, but the parameter was not changed.

The `hit_dir` parameter, as well as the code that modified it, was added in commit a3422b4d01.  Here is the relevant part:
```C++
@@ -236,7 +240,7 @@ void update_danger_weapon(object *ship_obj, object *weapon_obj)

 // function to actually deal with weapon-ship hit stuff.  separated from check_collision routine below
 // because of multiplayer reasons.
-void ship_weapon_do_hit_stuff(object *ship_obj, object *weapon_obj, vector *world_hitpos, vector *hitpos, int quadrant_num, int submodel_num)
+void ship_weapon_do_hit_stuff(object *ship_obj, object *weapon_obj, vector *world_hitpos, vector *hitpos, int quadrant_num, int submodel_num, vector /*not a pointer intentionaly*/ hit_dir)
 {
        weapon  *wp = &Weapons[weapon_obj->instance];
        weapon_info     *wip = &Weapon_info[wp->weapon_info_index];
@@ -287,13 +291,19 @@ void ship_weapon_do_hit_stuff(object *ship_obj, object *weapon_obj, vector *worl
        }

        if(quadrant_num == -1){
-               /*weapon_info   *wip = &Weapon_info[Weapons[weapon_obj->instance].weapon_info_index];
+               weapon_info     *wip = &Weapon_info[Weapons[weapon_obj->instance].weapon_info_index];
                decal_point dec;
                dec.orient = weapon_obj->orient;
+               vector hit_fvec;
+               vm_vec_negate(&hit_dir);
+               vm_vec_avg(&hit_fvec, &hit_dir, &weapon_obj->orient.vec.fvec);
+               vm_vec_normalize(&hit_fvec);
+               dec.orient.vec.fvec = hit_fvec;
+               vm_fix_matrix(&dec.orient);
                dec.pnt.xyz = hitpos->xyz;
                dec.radius = wip->decal_rad;
                if((dec.radius > 0) && (wip->decal_texture > -1))
-               decal_create(ship_obj, &dec, submodel_num, wip->decal_texture, wip->decal_backface_texture );*/
+               decal_create(ship_obj, &dec, submodel_num, wip->decal_texture, wip->decal_backface_texture, wip->decal_glow_texture, wip->decal_burn_texture, wip->decal_burn_time);
        }
```

Reply:
By "some kind of comment" I meant a *code* comment; if for no other reason than so that a future change won't accidentally reintroduce the reason it was copied in the first place.

Reply:
Why though?  The fact that the pointer is `const` will guard against the source vector being modified, which was the reason for the copy.  The code that actually *modified* the copy no longer exists.

Adding a comment reminds me of trying to prove a negative.  "If you ever add code to modify this vector, you'll need to make a local copy."  You could just as reasonably make that comment for *any* `const vec3d *` in the codebase.

Reply:
Not every `const vec3d *` in the codebase used to be a `vec3d` with a comment saying "not a pointer intentionaly".

Reply:
There wasn't a 200 IQ master plan behind that comment, just a very simple reason: the variable was copied because it was modified locally.  The previous code has been superseded; the current code no longer modifies it; and this PR enforces const-ness.

The original comment, and commit, was made by Bobboau.  He had, shall we say, a rather creative way of implementing many of his features.  There were many times when he pioneered some feature that was rewritten later on, to a greater or lesser extent (e.g. decals, fighter beams, model animations).  In none of those cases did we need to add comments explaining an API that was no longer in use.

If you won't budge, I'll add the comment, but I don't think it's either necessary or clarifying.

Reply:
> There wasn't a 200 IQ master plan behind that comment, just a very simple reason: the variable was copied because it was modified locally. The previous code has been superseded; the current code no longer modifies it; and this PR enforces const-ness.

The reason for the comment is irrelevant; the point is that anyone looking at the diff would see a comment being removed and the thing the comment was trying to prevent from happening actually happened. The commit message didn't explicitly explain, and only "alluded to" explaining it. Any confusion caused by this could be avoided by a very simple clarifying comment being added at the same time.

Reply:
That rationale makes a lot more sense.  In that case, I propose amending the commit message.  Would that be acceptable?

Reply:
Is there some reason not to comment the code, where the rationale would be immediately visible upon seeing the diff, rather than putting it in the commit message, where there would be an extra step to comprehension?

Reply:
Unless I'm mis-reading this discussion, I think you should take this suggestion to add a comment as a way to ease others into understanding why you made the change you made and why it is not a mistake.  Just because the code can be read by itself and understood, doesn't mean that adding a comment is valueless.  

You add value to code with comments by scaffolding other's understanding of it. You can potentially save someone a lot of investigative time whever something really subtle is wrong and every last line has to be analyzed.  We get plenty of bug reports that can't be reproduced, and you can't always rely on the debugger.

Reply:
> Is there some reason not to comment the code, where the rationale would be immediately visible upon seeing the diff, rather than putting it in the commit message, where there would be an extra step to comprehension?

A person viewing the diff would also be viewing the commit message, and therefore any confusion about the *change* in code can be clarified in the commit message.  But a person viewing the current state of the code would see only the current, correct version.

@JohnAFernandez I agree 100% with you in general, but this particular situation is an exception, for the reasons stated earlier.

Reply:
Well I have amended the previous commit to clarify the commit message, and added a new commit to comment unnecessarily about the `hit_dir` parameter.  I hope this is satisfactory.